cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -276,9 +286,22 @@ private long processResponse(Map<String, Deque<MonitorMessage>> monitorResponses
   private long publish(String topicConfig, String topic, Deque<MonitorMessage> messages,
                        AppMetadataStore store) throws Exception {
     // publish messages to tms
-    MessagePublisher messagePublisher = messagingContext.getMessagePublisher();
-    messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic,
-                             messages.stream().map(MonitorMessage::getMessage).iterator());
+    if (topic.startsWith(cConf.get(Constants.Logging.TMS_TOPIC_PREFIX))) {
+      messages.forEach(monitorMessage -> {
```

Comment:
It's better to move this logic into the `RemoteExecutionLogProcessor` implementation, especially for the Kafka case, we should't need to deserialize and reserialize the object.